File: app/src/main/java/de/bikebean/app/ui/utils/sms/listen/SmsListener.java
package de.bikebean.app.ui.utils.sms.listen;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.provider.Telephony;
import android.telephony.SmsMessage;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.preference.PreferenceManager;

import de.bikebean.app.ui.main.status.StateViewModel;
import de.bikebean.app.ui.main.status.menu.log.LogViewModel;
import de.bikebean.app.ui.main.status.menu.sms_history.SmsViewModel;

public class SmsListener extends BroadcastReceiver {
    /*
     SMS Listener

     This class listens for new SMS and calls
     our viewModel to update the SMS in background.
     */
    private static SmsViewModel sm;
    private static StateViewModel st;
    private static LogViewModel lv;

    public static void setViewModels(SmsViewModel smsViewModel, StateViewModel stateViewModel,
                                     LogViewModel logViewModel) {
        sm = smsViewModel;
        st = stateViewModel;
        lv = logViewModel;
    }

    @Override
    public void onReceive(@NonNull Context ctx, @NonNull Intent intent) {
        if (lv == null || st == null || sm == null)
            return;

        lv.d("Received something...");
        if (intent.getAction() == null
                || !intent.getAction().equals(Telephony.Sms.Intents.SMS_RECEIVED_ACTION))
            return;
        else
            lv.d("New SMS received...");

        final @Nullable String address = PreferenceManager.getDefaultSharedPreferences(ctx)
                .getString("number", null);
        if (address == null) {
            lv.e("Failed to load BB-number! Maybe it's not set?");
            return;
        }

        for (@NonNull SmsMessage smsMessage : Telephony.Sms.Intents.getMessagesFromIntent(intent)) {
            final @Nullable String senderAddress = smsMessage.getOriginatingAddress();

            if (senderAddress != null)
                if (senderAddress.equals(address)) {
                    lv.d("Sender is ours, start fetching sms");
                    sm.fetchSms(ctx, st, lv, address);
                } else
                    lv.d("Sender is not ours. (" + senderAddress + ")");
            else
                lv.w("Failed to get sender address!");
        }
    }
}

File: app/src/main/java/de/bikebean/app/ui/utils/sms/listen/SmsListener.kt
package de.bikebean.app.ui.utils.sms.listen

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.provider.Telephony
import de.bikebean.app.MainActivity
import de.bikebean.app.ui.utils.preferences.PreferencesUtils

class SmsListener : BroadcastReceiver() {

    /*
     SMS Listener

     This class listens for new SMS and calls
     back to MainActivity to handle them
     */
    companion object {
        @JvmStatic
        val newSmsString = "NEW_SMS"
    }

    override fun onReceive(ctx: Context, intent: Intent) {
        /*
         When a new message arrives, just trigger the MainActivity
         */

        // Exit if the intent is something else
        isIncomingSms(intent.action) ?: return

        (PreferencesUtils.getBikeBeanNumber(ctx) ?: return).let { number ->
            // Exit if the message is NOT from our bikeBean
            isFromOurBikeBean(intent, number) ?: return
        }

        // Trigger the MainActivity for a new message
        Intent(ctx, MainActivity::class.java)
                .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                .putExtra(newSmsString, 1)
                .let(ctx::startActivity)
    }

    private fun isIncomingSms(action: String?): Unit? = when {
        action != null && action == Telephony.Sms.Intents.SMS_RECEIVED_ACTION -> Unit
        else -> null
    }

    private fun isFromOurBikeBean(intent: Intent, number: String): Unit? = when {
        Telephony.Sms.Intents.getMessagesFromIntent(intent).any {
            it.originatingAddress != null && it.originatingAddress.equals(number)
        } -> Unit
        else -> null
    }
}